proxy/http2/Http2ServerSession.cc

```diff
@@ -0,0 +1,420 @@
+
+  this->add_session();
+}
+
+// implement that. After we send a GOAWAY, there
+// are scenarios where we would like to complete the outstanding streams.
+
+void
+Http2ServerSession::do_io_close(int alerrno)
+{
+  REMEMBER(NO_EVENT, this->recursion)
+
+  if (!this->connection_state.is_state_closed()) {
+    Http2SsnDebug("session closed");
+    this->remove_session();
+
+    ink_assert(this->mutex->thread_holding == this_ethread());
+    send_connection_event(&this->connection_state, HTTP2_SESSION_EVENT_FINI, this);
+
+    // Destroy will be called from connection_state.release_stream() once the number of active streams goes to 0
+  }
+}
+
+int
+Http2ServerSession::main_event_handler(int event, void *edata)
+{
+  ink_assert(this->mutex->thread_holding == this_ethread());
+  int retval;
+
+  recursion++;
+
+  Event *e = static_cast<Event *>(edata);
+  if (e == schedule_event) {
+    schedule_event = nullptr;
+  }
+
+  Http2SsnDebug("main_event_handler=%d edata=%p", event, edata);
+
+  switch (event) {
+  case VC_EVENT_READ_COMPLETE:
+  case VC_EVENT_READ_READY: {
+    bool is_zombie = connection_state.get_zombie_event() != nullptr;
+    retval         = (this->*session_handler)(event, edata);
+    if (is_zombie && connection_state.get_zombie_event() != nullptr) {
+      Warning("Processed read event for zombie session %" PRId64, connection_id());
+    }
+    break;
+  }
+
+  case HTTP2_SESSION_EVENT_REENABLE:
+    // VIO will be reenableed in this handler
+    retval = (this->*session_handler)(VC_EVENT_READ_READY, static_cast<VIO *>(e->cookie));
+    // Clear the event after calling session_handler to not reschedule REENABLE in it
+    this->_reenable_event = nullptr;
+    break;
+
+  case VC_EVENT_ACTIVE_TIMEOUT:
+  case VC_EVENT_INACTIVITY_TIMEOUT:
+  case VC_EVENT_ERROR:
+  case VC_EVENT_EOS:
+    this->set_dying_event(event);
+    this->do_io_close();
+    retval = 0;
+    break;
+
+  case VC_EVENT_WRITE_READY:
+  case VC_EVENT_WRITE_COMPLETE:
+    this->connection_state.restart_streams();
+    if ((Thread::get_hrtime() >= this->_write_buffer_last_flush + HRTIME_MSECONDS(this->_write_time_threshold))) {
+      this->flush();
+    }
+
+    retval = 0;
+    break;
+
+  case HTTP2_SESSION_EVENT_XMIT:
+  default:
+    Http2SsnDebug("unexpected event=%d edata=%p", event, edata);
+    ink_release_assert(0);
+    retval = 0;
+    break;
+  }
+
+  if (!this->is_draining() && this->connection_state.get_shutdown_reason() == Http2ErrorCode::HTTP2_ERROR_MAX) {
+    this->connection_state.set_shutdown_state(HTTP2_SHUTDOWN_NONE);
+  }
+
+  if (this->connection_state.get_shutdown_state() == HTTP2_SHUTDOWN_NONE) {
+    if (this->is_draining()) { // For a case we already checked Connection header and it didn't exist
+      Http2SsnDebug("Preparing for graceful shutdown because of draining state");
+      this->connection_state.set_shutdown_state(HTTP2_SHUTDOWN_NOT_INITIATED);
+    } /*else if (this->connection_state.get_stream_error_rate() >
```

Comment:
I don't think this part should be completely removed. If ATS is used as a forward proxy, it could be attacked by malicious/misbehaving origin servers.

There's no graceful shutdown mechanism defined on the spec, but we could stop making new requests on the connection that has high error rate to make the connection inactive and close it sooner rather than later.

We could also mark the origin server as a bad H2 server and use H1 on future connections. It should work as fallback mechanism. I think we need this kind of marking in the future anyway to choose a transport protocol (TCP or QUIC) for an origin server.

Reply:
But if we are going to say H2 to Origin is experimental support, like QUIC, we can ignore security concerns for now.

Reply:
Could add this back in.  As I dimly recall in my testing, the shutdown due to spurious errors seemed high. But could have been due to other reasons.